### ISSUE TITLE
Syntax highlighting inconsistencies

### ISSUE THREAD

**Issue by atnbueno** [openscad/openscad#2292]
I've been looking into updating the [Visual Studio Code extension for OpenSCAD](https://marketplace.visualstudio.com/items?itemName=efbenson.scad) because I've seen that some keywords were not highlighted. I took a list of keywords and tokens from the source code ([scadlexer.cpp](), [highlighter.cc]()), and a couple more things I found in the documentation, and tested using this file: [keywords_tokens.scad]()

[scadlexer.cpp]: https://github.com/openscad/openscad/blob/master/src/scadlexer.cpp
[highlighter.cc]: https://github.com/openscad/openscad/blob/master/src/highlighter.cc
[keywords_tokens.scad]: https://www.mundoplus.tv/atnbueno/keywords_tokens.scad

My surprise was when I checked it with the embedded editor as reference. I was surprised to see keywords like `intersection_for` and operators like `/` not highlighted either (captured below while using the *Tomorrow Night* theme):
![openscad_syntax_highlight_embedded_editor.png](https://mundoplus.tv/atnbueno/openscad_syntax_highlight_embedded_editor.png)

Specifically I see:
- `intersection_for` not highlighted
- Some operators not highlighted (at least division `/` and modulo `%`)
- Some deprecated functions highlighted (`child`) but others not (`assign`, `import_stl` and `import_dxf`)
- Single-quoted strings highlighted as valid although they generate errors
- Values `true`, `false` and `undef` highlighted the same as `module` and `function`

For comparison, this is what I'm doing (WIP!) with the Visual Studio Code extension:
![openscad_syntax_highlight_vscode_atnbueno.png](https://mundoplus.tv/atnbueno/openscad_syntax_highlight_vscode_atnbueno.png)

I'd love to know enough to fix this myself and submit a PR, but it's not the case :-P

P.S. Using Windows 10 x64 and version 2018.01.06

**Comment by MichaelPFrey** [comment]
With the operators: Try a more realistic syntax and it works:

![operators](https://user-images.githubusercontent.com/24962768/36113805-5492fb90-102e-11e8-8a22-02090e9397c7.png)

**Comment by MichaelPFrey** [comment]
scadlexer.cpp is used by scintillaeditor which is the default editor.
intersection_for can simply be added to the keyword list in scadlexer.cpp.

Question: Do you want to make the pull request? If not, I could do it.
_(Fixing part of the syntax highlighting was my first pull request here, see #2128 and I do not want to "steal" this fix from you, as we all start by picking low hanging fruits :-) )_

**Comment by atnbueno** [comment]
Nice of you to offer :-) but go ahead.

**Comment by MichaelPFrey** [comment]
Single-quoted strings:
**scintillaeditor.cpp**
```
		l->setColor(readColor(colors, "string", textColor), QsciLexerCPP::SingleQuotedString);
		l->setColor(readColor(colors, "string", textColor), QsciLexerCPP::DoubleQuotedString);
```
It is as simple as commenting out the SingleQuotedString line.

I was a bit confused - OpenSCAD is bypassing some logic of qscintilla for the colorScheme feature.

**Comment by MichaelPFrey** [comment]
deprecated functions: The more I think about it, it would make sense to make a new keyword set, so that we can give them a separate color/style.

**Comment by MichaelPFrey** [comment]
About adding a new keyword list:
I followed the trail into the rabbit hole and all I can comeback is a note:
LexCPP.cpp has only four keyword lists and even that is kind of bending their specific intended use.
Changing LexCPP.cpp is not really an option here down stream. Changing it upstream would take ages to tickle down.

I will think about adding a comment for the next person that tries to add one more keyword list.

**Comment by kintel** [comment]
@MichaelPFrey There was some work done to replace the Scintilla parser with something better, done as part of a GSoC project, but it needs someone to pick up the threads and finalize it. I seem to remember that the developer (Shaina) got pretty far though: http://brlcad.org/wiki/Google_Summer_of_Code/2015#ScadLexer_for_OpenSCAD

**Comment by MichaelPFrey** [comment]
Interesting, the source is in the branch [lexertl](https://github.com/openscad/openscad/tree/lexertl).
Well, that would be a bit to big for me in scope, at least at the moment.

**Comment by t-paul** [comment]
Closed via #4026.
